src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java

```diff
@@ -0,0 +1,197 @@
+        final long startTime = clock.millis();
+
+        getFileContentsForDirectory(storedRequestsDirectory)
+                .compose(storedIdToRequest -> getFileContentsForDirectory(storedImpressionsDirectory)
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(storedDataResult, startTime, MetricName.initialize))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.initialize));
+    }
+
+    private void refresh() {
+        final long startTime = clock.millis();
+
+        getFileContentsForDirectory(storedRequestsDirectory)
+                .compose(storedIdToRequest -> getFileContentsForDirectory(storedImpressionsDirectory)
+                        .map(storedIdToImp ->
+                                StoredDataResult.of(storedIdToRequest, storedIdToImp, Collections.emptyList())))
+                .onSuccess(storedDataResult -> handleResult(invalidate(storedDataResult), startTime, MetricName.update))
+                .onFailure(exception -> handleFailure(exception, startTime, MetricName.update));
+    }
+
+    private Void handleResult(StoredDataResult storedDataResult,
+                              long startTime,
+                              MetricName refreshType) {
+
+        lastResult.set(storedDataResult);
+
+        cacheNotificationListener.save(storedDataResult.getStoredIdToRequest(), storedDataResult.getStoredIdToImp());
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+
+        return null;
+    }
+
+    private Future<Void> handleFailure(Throwable exception, long startTime, MetricName refreshType) {
+        logger.warn("Error occurred while request to s3 refresh service", exception);
+
+        metrics.updateSettingsCacheRefreshTime(cacheType, refreshType, clock.millis() - startTime);
+        metrics.updateSettingsCacheRefreshErrorMetric(cacheType, refreshType);
+
+        return Future.failedFuture(exception);
+    }
+
+    private StoredDataResult invalidate(StoredDataResult storedDataResult) {
+        final Set<String> invalidatedRequests = getInvalidatedKeys(
+                storedDataResult.getStoredIdToRequest(),
+                lastResult != null ? lastResult.get().getStoredIdToRequest() : Collections.emptyMap());
+        final Set<String> invalidatedImps = getInvalidatedKeys(
+                storedDataResult.getStoredIdToImp(),
+                lastResult != null ? lastResult.get().getStoredIdToImp() : Collections.emptyMap());
+
+        if (!invalidatedRequests.isEmpty() || !invalidatedImps.isEmpty()) {
+            cacheNotificationListener.invalidate(
+                    invalidatedRequests.stream().toList(),
+                    invalidatedImps.stream().toList()
+            );
+        }
+
+        return storedDataResult;
+    }
```

Comment:
Why do we need this type of invalidation? This doesn't make any sense to me. Maybe you want to do invalidation like in `DatabasePeriodicRefreshService`? If so, consider re-implementing this code.

Reply:
I read the code and I assume my colleague adapted the code from the `DatabasePeriodicRefreshService` . 

Same as what the database service does, all ids that are not present in the current cache are passed to the cache listener to invalidate the cache. The difference seems to be that the database service invalidates all ids, while the S3 settings makes a diff. 

Maybe we could get rid of the `lastResult` atomic reference all together.

Unfortunately I can't even start prebid server locally anymore to see what happens :/ 

```
java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Pool closed
```

My run parameters are

```
--server.http.port=10000 --admin.port=10090 --metrics.prometheus.port=10091 --host-id=local --external-url=http://local.h5v.eu --spring.config.additional-location=sample/configs/prebid-config-s3.yaml
```

It works if I remove the `in-memory-cache` configuration.

Reply:
If so, then something went wrong when adapting the invalidation code 😅

Why I think so:
 - `DatabasePeriodicRefreshService ` invalidates all data that was retrieved as `null` or a blank string, which means that the data has been removed from the database. Because of this, we can't just put `null` or a blank string into the cache, we need to explicitly invalidate it.
 - Proposed implementation, in turn, for some reason tries to invalidate data that is in the new `fetch_result`, but is missing from the old one. Such behavior will most likely lead us to, at a minimum, meaningless attempts to invalidate data that isn't in the cache anyway, and, at most, will be "annoying" by invalidating data that can still be absolutely legally in the cache.

Reply:
That makes totally sense. Returning null or empty strings from the DB is the way then to invalidate the cache when used with the DB.

I guess I remove the entire invalidation then.

Reply:
Regarding `java.util.concurrent.CompletionException`... Can you share full stack trace, please?

Reply:
Sure thing. There's not a lot more info IMHO

```
java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Pool closed
	at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:64)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.lambda$execute$0(AsyncExecutionFailureExceptionReportingStage.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:78)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage2$RetryingExecutor.maybeAttemptExecute(AsyncRetryableStage2.java:135)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage2$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage2.java:152)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage2$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage2.java:113)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:78)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$execute$0(MakeAsyncHttpRequestStage.java:108)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.completeResponseFuture(MakeAsyncHttpRequestStage.java:255)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$5(MakeAsyncHttpRequestStage.java:184)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$6(MakeAsyncHttpRequestStage.java:183)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
	at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.onError(AsyncResponseHandler.java:79)
	at software.amazon.awssdk.core.internal.http.async.AsyncAfterTransmissionInterceptorCallingResponseHandler.onError(AsyncAfterTransmissionInterceptorCallingResponseHandler.java:72)
	at software.amazon.awssdk.core.internal.http.async.FilterTransformingAsyncHttpResponseHandler.onError(FilterTransformingAsyncHttpResponseHandler.java:54)
	at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.handleFailure(NettyRequestExecutor.java:308)
	at software.amazon.awssdk.http.nio.netty.internal.NettyRequestExecutor.makeRequestListener(NettyRequestExecutor.java:188)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:557)
	at io.netty.util.concurrent.DefaultPromise.access$200(DefaultPromise.java:35)
	at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:503)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask$$$capture(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Pool closed
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:47)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper2.setLastException(RetryableStageHelper2.java:226)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper2.setLastException(RetryableStageHelper2.java:221)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage2$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage2.java:151)
	... 43 common frames omitted
Caused by: java.lang.IllegalStateException: Pool closed
	at software.amazon.awssdk.http.nio.netty.internal.http2.HttpOrHttp2ChannelPool.lambda$completeProtocolConfiguration$5(HttpOrHttp2ChannelPool.java:141)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:106)
	... 9 common frames omitted
```

Reply:
@CTMBNara I'm not getting any closer in solving this.

I have tried

- [explicitly configuring netty](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/http-configuration-netty.html) - adding only one connection aquire at a time, longer timeout, disabling TLS as hinted here https://github.com/aws/aws-sdk-java-v2/issues/1766
- Using https://github.com/reactiverse/aws-sdk , the error went away, but the application just crashed with a "NonStrackTrace: null" exception. I assume because I added `Vertx vertx` as a dependency to the `S3AsyncClient` bean constructor
- removing the `vert.getOrCreateContext` call in the `S3PeriodicRefreshService`


Here's the configuration

```
        @Bean
        S3AsyncClient s3AsyncClient(S3ConfigurationProperties s3ConfigurationProperties) throws URISyntaxException {
            final AwsBasicCredentials credentials = AwsBasicCredentials.create(
                    s3ConfigurationProperties.getAccessKeyId(),
                    s3ConfigurationProperties.getSecretAccessKey());
            final String awsRegionName = s3ConfigurationProperties.getRegion();
            final Region awsRegion = awsRegionName != null
                    ? Region.of(s3ConfigurationProperties.getRegion())
                    : Region.AWS_GLOBAL;
            return S3AsyncClient
                    .builder()
                    .credentialsProvider(StaticCredentialsProvider.create(credentials))
                    .endpointOverride(new URI(s3ConfigurationProperties.getEndpoint()))
                    .region(awsRegion)
// reactiverse/aws-sdk trial
//                    .httpClient(
//                            new VertxNioAsyncHttpClient(vertx.getOrCreateContext()) // uses Vert.x's HttpClient to make call to AWS services
//                    )
//                    .asyncConfiguration(conf -> // tells AWS to execute response callbacks in a Vert.x context
//                            conf.advancedOption(SdkAdvancedAsyncClientOption.FUTURE_COMPLETION_EXECUTOR,
//                                    new VertxExecutor(vertx.getOrCreateContext()))
//                    )
                    .httpClient(software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient.builder()
                            .maxPendingConnectionAcquires(1)
                            .connectionTimeout(java.time.Duration.ofSeconds(20))
                            .connectionAcquisitionTimeout(java.time.Duration.ofSeconds(20))
                            .connectionMaxIdleTime(java.time.Duration.ofSeconds(5))
                            .buildWithDefaults(AttributeMap.builder()
                                    .put(SdkHttpConfigurationOption.TLS_KEY_MANAGERS_PROVIDER, TlsKeyManagersProvider.noneProvider())
                                    .put(SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES, true)
                                    .build()
                            )

                    )
                    .build();
        }
```

My assumption is, that there's some sort of race condition or wrong order of initialization. Because when I disable the `in-memory-cash`, then every works. Startup, requests, etc.

```
settings:
  enforce-valid-account: false
  generate-storedrequest-bidrequest-id: true
  s3:
    accessKeyId: prebid-server-test
    secretAccessKey: much-secret-very-wow
    endpoint: http://localhost:9000
    bucket: prebid-server-configs-dev.h5v.eu # prebid-application-settings
    # region: <region name> # if not provided AWS_GLOBAL will be used. Example value: 'eu-central-1'
    accounts-dir: accounts
    stored-imps-dir: stored-impressions
    stored-requests-dir: stored-requests
    stored-responses-dir: stored-responses

# with this enabled, the server crashes on startup
#  in-memory-cache:
#    cache-size: 10000
#    ttl-seconds: 1200 # 20 minutes
#    s3-update:
#      refresh-rate: 900000 # Refresh every 15 minutes
#      timeout: 5000
```

Reply:
When turning on debug logging for `netty` I see that the `CurrencyConversionService` opens a connection to load the latest.json from jsdelvr. Maybe they share some internal pool from netty, which is killed by the currency conversion service after initialization.

And I found this sneaky line

```
2024-08-17T13:53:34.050+02:00  WARN 72949 --- [ntloop-thread-1] o.p.s.c.CurrencyConversionService        : Error occurred while request to currency service
```

Reply:
Hi @muuki88 and sorry I was in vacation :) Yeah, you are right: we need to remove `vertx.getOrCreateContext`, but ONLY from `S3PeriodicRefreshService`, because all the methods already called inside vertx context.

Reply:
Done :smile: 

And no worries about. We will get this through at some point :joy:

Reply:
@CTMBNara the startup error [described above still remains though](https://github.com/prebid/prebid-server-java/pull/2733#discussion_r1720763960). 

I'm 80% certain that there's a hidden bug, that slept because there was no other library using netty as an http client.

Reply:
@muuki88 Try this code:
```java
    @Override
    public void initialize(Promise<Void> initializePromise) {
        final long startTime = clock.millis();

        fetchStoredDataResult()
                .onSuccess(storedDataResult -> handleResult(storedDataResult, startTime, MetricName.initialize))
                .onFailure(exception -> handleFailure(exception, startTime, MetricName.initialize));
        if (refreshPeriod > 0) {
            vertx.setPeriodic(refreshPeriod, ignored -> refresh());
        }

        initializePromise.tryComplete();
    }
```

Reply:
This works, but it's kinda strange, because it just swallows the exception.

The point of this `initializePromise` is to check if that service was started properly. If it can't start, then there's no point in starting up prebid server at all, because all configuration is missing.

While this "fixes" the issue, it allows prebid server to start up in a useless state.

WDYT?

Reply:
Yes, that's how it works) 
But it's intentional idea to all services to not fail the application if data is not required for runtime work.
For now, please disregard this point.

Reply:
Considering the above, please use the following [implementation](https://github.com/prebid/prebid-server-java/pull/2733#discussion_r1723338114)

Reply:
Okay, so back to reverting the reverts and ignoring this comment?
https://github.com/prebid/prebid-server-java/pull/2733#issuecomment-2299529107

> Yeah, you are right. Thanks that you accidentally found it :) Bug-fix: https://github.com/prebid/prebid-server-java/pull/3385

Reply:
@muuki88 there is no need for ignoring and reverts.